NPE in BaseHoodieTableFileIndex.getInputFileSlices when RO path filter is enabled and a queried partition has no base files

## Bug Description

**What happened:**

When `hoodie.datasource.read.file.index.list.file.statuses.using.ro.path.filter=true` is set on a COW table (or on any table queried with `READ_OPTIMIZED`), querying a partition set that includes one or more partitions with no base files throws `NullPointerException` deep inside Spark query planning, originating from `BaseHoodieTableFileIndex.getInputFileSlices`.

**What you expected:**

The query should plan and execute successfully. Empty partitions should appear in the result map with an empty file-slice list, matching the contract already honored by the non-RO path (`filterFiles`), which iterates over partitions and naturally returns an entry per partition.

**Steps to reproduce:**

1. Create a COW table with multiple partitions, where at least one partition contains no base files (e.g., produced by clustering, cleaning, or simply never written).
2. Enable `hoodie.datasource.read.file.index.list.file.statuses.using.ro.path.filter=true`.
3. Issue a Spark SQL query that prunes down to a partition set including the empty one.
4. The driver throws `NullPointerException` during query optimization.

**Root cause:**

`BaseHoodieTableFileIndex#loadFileSlicesForPartitions` has two code paths:

| Path | Method | Behavior for empty partitions |
|------|--------|-------------------------------|
| A | `generatePartitionFileSlicesPostROTablePathFilter` (added in #18136) | iterates over **files**, so a partition with zero files gets **no entry** in the returned map |
| B | `filterFiles` | iterates over **partitions**, so every partition gets an entry (possibly empty list) |

When Path A is taken, `cachedAllInputFileSlices` is left without entries for empty partitions. The caller `getInputFileSlices` then does:

```java
return Arrays.stream(partitions).collect(
    Collectors.toMap(Function.identity(), partition -> cachedAllInputFileSlices.get(partition))
);
```

`cache.get(emptyPartition)` returns `null`, and `Collectors.toMap`'s `uniqKeysMapAccumulator` calls `Objects.requireNonNull(value)` → NPE.

A fix and regression test are in PR #<PR_NUMBER>.

## Environment

**Hudi version:** master (bug introduced by #18136 in 1.x branch; affects 1.2 and later releases that include the optimization)
**Query engine:** Spark 3.3.x (also affects other Spark versions; bug is in `hudi-common`)
**Relevant configs:**
- `hoodie.datasource.read.file.index.list.file.statuses.using.ro.path.filter=true` (the trigger; default is `false`)
- Table type `COPY_ON_WRITE` *or* query type `READ_OPTIMIZED` (either condition activates Path A)

## Logs and Stack Trace

```
Caused by: java.lang.NullPointerException
    at java.util.Objects.requireNonNull(Objects.java:222)
    at java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:178)
    at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
    at org.apache.hudi.BaseHoodieTableFileIndex.getInputFileSlices(BaseHoodieTableFileIndex.java:252)
    at org.apache.hudi.HoodieFileIndex.prunePartitionsAndGetFileSlices(HoodieFileIndex.scala:351)
    at org.apache.hudi.HoodieFileIndex.filterFileSlices(HoodieFileIndex.scala:227)
    at org.apache.spark.sql.hudi.analysis.Spark33HoodiePruneFileSourcePartitions$$anonfun$apply$1.applyOrElse(Spark33HoodiePruneFileSourcePartitions.scala:63)
    ... (Spark catalyst frames)
```

--- a/hudi-common/src/main/java/org/apache/hudi/BaseHoodieTableFileIndex.java
+++ b/hudi-common/src/main/java/org/apache/hudi/BaseHoodieTableFileIndex.java
@@ -314,6 +314,11 @@ public abstract class BaseHoodieTableFileIndex implements AutoCloseable {
     Map<String, PartitionPath> partitionsMap = new HashMap<>();
     partitions.forEach(p -> partitionsMap.put(p.path, p));
     Map<PartitionPath, List<FileSlice>> partitionToFileSlices = new HashMap<>();
+    // Pre-populate so partitions with no files still appear in the result map.
+    // Without this, the caller's Collectors.toMap(identity, cache::get) NPEs on empty partitions
+    // because cache.get returns null and toMap rejects null values. This matches the contract
+    // already honored by filterFiles, which iterates over partitions rather than over files.
+    partitions.forEach(p -> partitionToFileSlices.put(p, Collections.emptyList()));
 
     for (StoragePathInfo pathInfo : allFiles) {
       // Create FileSlice obj from StoragePathInfo.
@@ -326,7 +331,11 @@ public abstract class BaseHoodieTableFileIndex implements AutoCloseable {
       // Add the FileSlice to partitionToFileSlices
       PartitionPath partitionPathObj = partitionsMap.get(relPartitionPath);
       if (partitionPathObj != null) {
-        List<FileSlice> fileSlices = partitionToFileSlices.computeIfAbsent(partitionPathObj, k -> new ArrayList<>());
+        List<FileSlice> fileSlices = partitionToFileSlices.get(partitionPathObj);
+        if (fileSlices.isEmpty()) {
+          fileSlices = new ArrayList<>();
+          partitionToFileSlices.put(partitionPathObj, fileSlices);
+        }
         fileSlices.add(fileSlice);
       } else {
         log.warn("Could not find partition path object for relative path: {}. Skipping file: {}",

--- a/hudi-common/src/test/java/org/apache/hudi/BaseHoodieTableFileIndexTest.java
+++ b/hudi-common/src/test/java/org/apache/hudi/BaseHoodieTableFileIndexTest.java
@@ -18,15 +18,25 @@
 
 package org.apache.hudi;
 
+import org.apache.hudi.BaseHoodieTableFileIndex.PartitionPath;
 import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.model.FileSlice;
+import org.apache.hudi.storage.StoragePath;
+import org.apache.hudi.storage.StoragePathInfo;
 
 import org.junit.jupiter.api.Test;
 
 import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.Mockito.mock;
 
 public class BaseHoodieTableFileIndexTest {
@@ -57,5 +67,64 @@ public class BaseHoodieTableFileIndexTest {
     assertEquals(true, result.isEnabled(), "Metadata should be enabled");
     assertEquals(true, result.isBloomFilterIndexEnabled(), "Bloom filter index should be enabled");
     assertEquals(true, result.isColumnStatsIndexEnabled(), "Column stats index should be enabled");
+  }
+
+  /**
+   * Regression test for the empty-partition NPE that surfaces in {@code getInputFileSlices}
+   * when the {@code hoodie.datasource.read.file.index.list.file.statuses.using.ro.path.filter}
+   * code path is exercised on a COW (or READ_OPTIMIZED) table that contains a partition
+   * holding zero base files.
+   *
+   * <p>Before the fix, {@link BaseHoodieTableFileIndex#generatePartitionFileSlicesPostROTablePathFilter}
+   * built its result map by iterating over the file list, so a partition with no files received
+   * no entry. The downstream {@code Collectors.toMap(identity, p -> cache.get(p))} in
+   * {@code getInputFileSlices} then dereferenced a null value and threw NPE inside
+   * {@code Collectors.uniqKeysMapAccumulator}.
+   *
+   * <p>After the fix, every input partition appears in the returned map (with an empty list
+   * for empty partitions), preserving the contract already honored by the non-RO path
+   * ({@code filterFiles}).
+   */
+  @Test
+  public void testGeneratePartitionFileSlicesPostROTablePathFilterIncludesEmptyPartitions() throws Exception {
+    BaseHoodieTableFileIndex fileIndex = mock(BaseHoodieTableFileIndex.class,
+        org.mockito.Mockito.CALLS_REAL_METHODS);
+
+    StoragePath basePath = new StoragePath("/tmp/hudi_empty_partition_test");
+    Field basePathField = BaseHoodieTableFileIndex.class.getDeclaredField("basePath");
+    basePathField.setAccessible(true);
+    basePathField.set(fileIndex, basePath);
+
+    PartitionPath partitionWithFiles = new PartitionPath("dt=2026-01-01", new Object[]{"2026-01-01"});
+    PartitionPath emptyPartition = new PartitionPath("dt=2026-01-02", new Object[]{"2026-01-02"});
+    PartitionPath anotherEmpty = new PartitionPath("dt=2026-01-03", new Object[]{"2026-01-03"});
+    List<PartitionPath> partitions = Arrays.asList(partitionWithFiles, emptyPartition, anotherEmpty);
+
+    StoragePathInfo file = new StoragePathInfo(
+        new StoragePath(basePath, "dt=2026-01-01/file-0_0-0-0_20260101000000001.parquet"),
+        100L, false, (short) 1, 1024L, 0L);
+    List<StoragePathInfo> allFiles = Collections.singletonList(file);
+
+    Method generateMethod = BaseHoodieTableFileIndex.class.getDeclaredMethod(
+        "generatePartitionFileSlicesPostROTablePathFilter", List.class, List.class);
+    generateMethod.setAccessible(true);
+    @SuppressWarnings("unchecked")
+    Map<PartitionPath, List<FileSlice>> result =
+        (Map<PartitionPath, List<FileSlice>>) generateMethod.invoke(fileIndex, partitions, allFiles);
+
+    assertNotNull(result, "Result map must not be null");
+    assertEquals(3, result.size(),
+        "Result map must contain an entry for every input partition, including empty ones");
+    assertTrue(result.containsKey(partitionWithFiles));
+    assertTrue(result.containsKey(emptyPartition),
+        "Empty partition must appear in the result so getInputFileSlices does not NPE");
+    assertTrue(result.containsKey(anotherEmpty),
+        "Empty partition must appear in the result so getInputFileSlices does not NPE");
+    assertEquals(1, result.get(partitionWithFiles).size(),
+        "Partition with files should retain its file slice");
+    assertTrue(result.get(emptyPartition).isEmpty(),
+        "Empty partition's file slice list must be present and empty (not null, not missing)");
+    assertTrue(result.get(anotherEmpty).isEmpty(),
+        "Empty partition's file slice list must be present and empty (not null, not missing)");
   }
 }

